edgedb/_testbase.py

```diff
@@ -52,54 +54,74 @@ def flt(log_record):
 _default_cluster = None
 
 
-def _init_cluster(data_dir=None, *, cleanup_atexit=True):
-    if (not os.environ.get('EDGEDB_DEBUG_SERVER') and
-            not os.environ.get('EDGEDB_LOG_LEVEL')):
-        _env = {'EDGEDB_LOG_LEVEL': 'silent'}
-    else:
-        _env = {}
+def _start_cluster(*, cleanup_atexit=True):
+    global _default_cluster
 
-    if data_dir is None:
-        cluster = edgedb_cluster.TempCluster(env=_env, testmode=True)
-        destroy = True
-    else:
-        cluster = edgedb_cluster.Cluster(data_dir=data_dir, env=_env)
-        destroy = False
+    if isinstance(_default_cluster, Exception):
+        # when starting a server fails
+        # don't retry starting one for every TestCase
+        # because that takes a long time
+        raise _default_cluster
 
-    if cluster.get_status() == 'not-initialized':
-        cluster.init()
+    if _default_cluster:
+        return _default_cluster
 
-    cluster.start(port='dynamic')
-    cluster.set_superuser_password('test')
+    try:
+        status_file = pathlib.Path('/tmp/edgedb-server-status')
+        if sys.platform == 'win32':
+            status_file = pathlib.Path('/mnt/c/temp/edgedb-server-status.txt')
+
+        args = [
+            os.environ.get('EDGEDB_SERVER_BINARY', 'edgedb-server'),
+            "--temp-dir",
+            "--testmode",
+            f"--emit-server-status={status_file}",
+            "--port=auto",
+            "--bootstrap-command=ALTER ROLE edgedb { SET password := 'test' }",
+        ]
+        cwd = '/tmp'
+
+        if sys.platform == 'win32':
+            args = ['wsl', 'sudo', '-u', 'edgedb'] + args
+            cwd = 'C:\\Temp'
+
+        env = os.environ.copy()
+        # Make sure the PYTHONPATH of _this_ process does
+        # not interfere with the server's.
+        env.pop('PYTHONPATH', None)
+
+        p = subprocess.Popen(args, env=env, cwd=cwd)
+
+        if sys.platform == 'win32':
+            status_file = pathlib.Path(r'C:\Temp\edgedb-server-status.txt')
+
+        for i in range(2500):
+            print('iteration number:', i)
+            if status_file.exists():
+                break
+            time.sleep(1)
+        else:
+            raise RuntimeError('server status file not found')
 
-    if cleanup_atexit:
-        atexit.register(_shutdown_cluster, cluster, destroy=destroy)
+        with status_file.open('rb') as f:
```

Comment:
I'd put the reading of the file inside the retry loop as well, because otherwise you might get a race between the file getting created and data being flushed into it.  Basically: 1) try opening the file; 2) iterate over lines and look for `READY=`; 3) if found, break.

Reply:
changed